cmd/erasure-healfile.go

```diff
@@ -16,72 +16,136 @@
 
 package cmd
 
-import "hash"
+import (
+	"hash"
+)
+
+// HealFile tries to reconstruct a bitrot encoded file spread over all
+// available disks. HealFile will read the valid parts of the file,
+// reconstruct the missing data and write the reconstructed parts back
+// to `staleDisks`.
+//
+// `staleDisks` is a slice of disks where each non-nil entry
+// represents an online disk that is known to have stale or no data,
+// and so will be healed.
+//
+// The function will try to read the valid parts from the file under
+// the given volume and path and tries to reconstruct the file under
+// the given healVolume and healPath (on staleDisks). The given
+// algorithm will be used to verify the valid parts and to protect the
+// reconstructed file.
+//
+// It returns bitrot checksums for the non-nil staleDisks.
+func (s ErasureStorage) HealFile(staleDisks []StorageAPI, volume, path string,
+	blocksize int64, healVolume, healPath string, size int64,
+	algorithm BitrotAlgorithm, checksums [][]byte) (f ErasureFileInfo,
+	err error) {
 
-// HealFile tries to reconstruct a bitrot encoded file spread over all available disks. HealFile will read the valid parts of the file,
-// reconstruct the missing data and write the reconstructed parts back to the disks.
-// It will try to read the valid parts from the file under the given volume and path and tries to reconstruct the file under the given
-// healVolume and healPath. The given algorithm will be used to verify the valid parts and to protect the reconstructed file.
-func (s ErasureStorage) HealFile(offlineDisks []StorageAPI, volume, path string, blocksize int64, healVolume, healPath string, size int64, algorithm BitrotAlgorithm, checksums [][]byte) (f ErasureFileInfo, err error) {
 	if !algorithm.Available() {
 		return f, traceError(errBitrotHashAlgoInvalid)
 	}
+
+	// Initialization
 	f.Checksums = make([][]byte, len(s.disks))
-	hashers, verifiers := make([]hash.Hash, len(s.disks)), make([]*BitrotVerifier, len(s.disks))
+	hashers := make([]hash.Hash, len(s.disks))
+	verifiers := make([]*BitrotVerifier, len(s.disks))
 	for i, disk := range s.disks {
-		if disk == OfflineDisk {
+		switch {
+		case staleDisks[i] != nil:
```

Comment:
Those two cases are identical?! `staleDisk != nil` whenever `s.disk == OfflineDisk` and vice versa. Further we agreed on using `OfflineDisk` as indicator whether disk is available -> s/`staleDisk != nil`/`staleDisk != OfflineDisk`

Reply:
The slice of disks mean different things - `staleDisks` is non-nil when the disk is known to have out-of-date data. It is better to not conflate offline with nil.